### PR TITLE
external: S3 concurrent reader releases memory after it's not used

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #45719

Problem Summary:

### What is changed and how it works?

When turn off concurrent reader, free the memory immediately.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

run `TestMemoryUsageWhenHotspotChange`
before
```
    iter_test.go:486: memory usage: 10518528 -> 7654735872
```
after
```
    iter_test.go:493: memory usage: 10706944 -> 944136192
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/byte_reader.go [hunk 1]
```diff
@@ -17,33 +17,50 @@ package external
...
 	"github.com/pingcap/tidb/util/logutil"
 	"github.com/pingcap/tidb/util/mathutil"
 	"go.uber.org/zap"
 )
 
-// ConcurrentReaderBufferSize is the buffer size for concurrent reader.
-var ConcurrentReaderBufferSize = 4 * 1024 * 1024
+var (
+	// ConcurrentReaderBufferSize is the buffer size for concurrent reader.
+	ConcurrentReaderBufferSize = 4 * 1024 * 1024
```
Comment (Benjamin2037) [inline br/pkg/lightning/backend/external/byte_reader.go:30]: Is this mean reader buffer size 4M?

File: br/pkg/lightning/backend/external/byte_reader.go [hunk 2]
```diff
@@ -63,69 +80,118 @@ func openStoreReaderAndSeek(
...
+	}
+	// need to set it before reload()
+	readerFields.expected = useConcurrent
+	// concurrent reader will be lazily initialized when reload()
+	if useConcurrent {
+		return nil
+	}
+
+	// no change
+	if !readerFields.now && !useConcurrent {
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/byte_reader.go:138]: `!useConcurrent` always is true.

File: br/pkg/lightning/backend/external/byte_reader.go (hunk 1 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/byte_reader.go:30]: it's per concurrency of a reader, I'll rename it.

File: br/pkg/lightning/backend/external/byte_reader.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/byte_reader.go]: https://github.com/pingcap/tidb/pull/47239/commits/e8773f5304fd46e87e21a51ff8bacb28b440e4e5

File: br/pkg/lightning/backend/external/byte_reader.go (hunk 1 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/byte_reader.go]: https://github.com/pingcap/tidb/pull/47239/commits/e8773f5304fd46e87e21a51ff8bacb28b440e4e5

File: br/pkg/lightning/backend/external/byte_reader.go [hunk 3]
```diff
@@ -180,52 +247,70 @@ func (r *byteReader) cloneSlices() {
...
-	r.bufOffset = 0
+	r.curBufOffset = 0
 	return nil
 }
 
+func (r *byteReader) closeConcurrentReader() (reloadCnt, offsetInOldBuffer int) {
+	r.logger.Info("drop data in closeConcurrentReader",
+		zap.Int("reloadCnt", r.concurrentReader.reloadCnt),
+		zap.Int("dropBytes", len(r.curBuf)-r.curBufOffset),
+	)
```
[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/backend/external/byte_reader.go:299]: should be same amount like https://github.com/pingcap/tidb/blob/789adfcf9d46afb0cbc822ecd14733589e0a353c/br/pkg/lightning/backend/external/byte_reader.go#L196

File: br/pkg/storage/local.go [hunk 4]
```diff
@@ -151,27 +150,53 @@ func (l *LocalStorage) Open(_ context.Context, path string, o *ReaderOption) (Ex
...
+func (f *localFile) Read(p []byte) (n int, err error) {
+	if f.endPos == -1 {
+		return f.File.Read(p)
+	}
+
+	pEnd := f.endPos - f.pos
+	if pEnd <= 0 {
+		return 0, io.EOF
+	}
+	p = p[:pEnd]
```
Comment (tangenta) [inline br/pkg/storage/local.go:184]: Is this possible that `pEnd > len(p)`?

File: br/pkg/lightning/backend/external/kv_reader.go [hunk 5]
```diff
@@ -82,5 +82,6 @@ func noEOF(err error) error {
 }
 
 func (r *kvReader) Close() error {
+	r.byteReader.concurrentReader.largeBufferPool.Destroy()
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/kv_reader.go:85]: `largeBufferPool` is `nil` if the reader does not enable concurrent mode.

File: br/pkg/lightning/backend/external/iter.go [hunk 6]
```diff
@@ -205,20 +221,49 @@ func (i *mergeIter[T, R]) next() bool {
...
+			for idx, rp := range i.readers {
+				if rp == nil {
+					continue
+				}
+				isHotspot := i.lastHotspotIdx == idx
+				err := (*rp).switchConcurrentMode(isHotspot)
+				if err != nil {
+					i.err = err
+					return false
+				}
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/iter.go:253]: We only need to swtich at most two readers. Is there a plan to support concurrent read on multiple readers?

**Review by tangenta — COMMENTED** [review]
Rest LGTM

File: br/pkg/lightning/backend/external/kv_reader.go (hunk 5 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/kv_reader.go]: https://github.com/pingcap/tidb/pull/47239/commits/90f68418d25be240ed2db79aad4bd33f488c5205

File: br/pkg/lightning/backend/external/iter.go (hunk 6 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/iter.go:253]: Yes I hope so. After we log the dropped data we will can better evaluate current strategy, and see if it's needed to learn more about external merge sort prefetch

**Review by wjhuang2016 — APPROVED** [review]

**Review by Leavrth — APPROVED** [review]

**Pull request merged**